### PR TITLE
Consider reflogs of remote tracking branches in layout discovery as well

### PR DESCRIPTION
**Pull request by PawelLipski**

### CODE REVIEW

File: backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepository.java [hunk 1]
```diff
@@ -242,27 +245,28 @@ public IGitMacheteRepositorySnapshot discoverLayoutAndCreateSnapshot() throws Gi
...
+            Seq<ILocalBranchReference> eligibleContainingBranches = deriveBranchesContainingGivenCommitInReflog()
                 .getOrElse(commit.getHash(), List.empty())
-                .filter(candidateBranchName -> !candidateBranchName.equals(branch.getName())
-                    && !candidateBranchName.equals(remoteTrackingBranchName)
-                    && eligibleBranchNames.contains(candidateBranchName));
+                .map(candidateBranch -> candidateBranch.isLocal()
+                    ? candidateBranch.asLocal()
+                    : candidateBranch.asRemote().getTrackedLocalBranch())
+                .filter(correspondingLocalBranch -> !correspondingLocalBranch.getName().equals(localBranch.getName())
+                    && eligibleLocalBranchNames.contains(correspondingLocalBranch.getName()));
```
Comment (PawelLipski) [inline backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepository.java:269]: This is the actual fix

Comment (PawelLipski) [reply backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepository.java:269]: If a commit (ancestor of commit point by branch X) has been found only in reflog of origin/Y but not Y itself, let's still consider Y as a viable parent for X

File: backendImpl/src/test/java/com/virtuslab/gitmachete/backend/integration/StatusAndDiscoverIntegrationTestSuite.java [hunk 2]
```diff
@@ -183,14 +183,8 @@ private void printNonRootBranch(
       sb.append(c.getShortMessage());
       if (c.equals(forkPoint)) {
         sb.append(" -> fork point ??? commit ${forkPoint.getShortHash()} has been found in reflog of ");
-        Set<String> branchesContainingInReflog = forkPoint.getBranchesContainingInReflog().toSet();
-        // This is a hack necessary to keep consistent with CLI, which skips remote branch from the fork point hint
```
Comment (PawelLipski) [inline backendImpl/src/test/java/com/virtuslab/gitmachete/backend/integration/StatusAndDiscoverIntegrationTestSuite.java:187]: Yeah, a hack that backfired ;p

File: qual/src/main/java/com/virtuslab/qual/internal/SubtypingBottom.java [hunk 3]
```diff
@@ -20,11 +22,16 @@
 // There needs to be single subtyping hierarchy with single bottom and top annotation,
 // otherwise Subtyping Checker would raise an error about multiple top/bottom types.
 @SubtypeOf({
-    // Despite having a unified type hierarchy, we're actually doing 3 completely independent checks here.
+    // Despite having a unified type hierarchy, we're actually doing 4 completely independent checks here.
```
Comment (mkondratek) [inline qual/src/main/java/com/virtuslab/qual/internal/SubtypingBottom.java:25]: `an unified` :question:

Comment (mkondratek) [reply qual/src/main/java/com/virtuslab/qual/internal/SubtypingBottom.java:25]: N/A, you are right

Comment (PawelLipski) [reply qual/src/main/java/com/virtuslab/qual/internal/SubtypingBottom.java:25]: Yeah, `ˈjuːnɪfaɪd` ;)

**Pull request merged**